docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,663 @@
+
+{% comment %}{% include example.md title="A schema that imports types from other schemas" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+### Schema Authorities
+
+The structure of a `id` string is defined by the schema authority responsible for the schema/type(s) being imported.
+Note that runtime resolution of a schema over a network presents availability and security risks, and should therefore be avoided.
+
+When resolving a schema, authorities may choose to follow well-known patterns; for example:
+- a filesystem authority might specify that an `id` string corresponds to an ISL file relative to some base, e.g.:`"{base}/com/example/core/Customer.isl"`
+- a REST authority might specify that an `id` string is a resource URL that corresponds to an ISL file, e.g.: `"https://{host}:{port}/{base}/com/example/core/Customer"` (again, note the inherent availability and security risks here)
+
+Beware that the choice of authority affects the way schemas can be imported by other schemas.
+
+# Types
+
+The Ion Schema type system is a hybrid of [nominal](https://en.wikipedia.org/wiki/Nominal_type_system) and [structural](https://en.wikipedia.org/wiki/Structural_type_system) typing.
+All Ion values are nominally one of the types defined in the [Ion data model](https://amzn.github.io/ion-docs/docs/spec.html#the-ion-data-model).
+Beyond that, a value (or stream of values) may belong to one or more structural types.
+(All values belong to at least one structural type—the top type `$any`.)
+
+## Built-in Types
+
+Ion Schema has several built-in types that are implicitly imported before any other imports are handled.
+
+The nominal Ion types are prefixed with `$`, and correspond precisely with the types defined by the Ion data model, including strongly-typed null values:
+- **scalars:** `$blob`, `$bool`, `$clob`, `$decimal`, `$float`, `$int`, `$null`,
+  `$string`, `$symbol`, `$timestamp`
+- **containers:** `$list`, `$sexp`, `$struct`
+
+Ion Schema adds one more nominal type, `document`, which is a stream of top-level Ion values.
+Ion Schema implementations may represent this stream as a list, sequence, array, or other similar data structure depending on what is idiomatic for the language in which it is implemented.
+
+In addition, Ion Schema provides the following structural types:
+
+- `nothing`: the [empty type](https://en.wikipedia.org/wiki/Empty_type); it has no valid value
+- `$any`: the [top type](https://en.wikipedia.org/wiki/Top_type); every value is valid
+- `any`: represents any value that is not a null value
+- `blob`, `bool`, `clob`, `decimal`, `float`, `int`, `string`, `symbol`, `timestamp`, `list`, `sexp`, `struct` correspond to the Ion types, except that they do not allow null values.
+- `$lob`: represents a `$blob` or `$clob`
+- `lob`: represents a `blob` or `clob`
+- `$number`: represents a `$decimal`, `$float`, or `$int`
+- `number`: represents a `decimal`, `float`, or `int`
```

Comment:
More broadly... I read the writeup about the potential ambiguities with `nullable` in 1.0 and my recollection from working with it was that it was confusing. So I agree that we can and should do better with 2.0.

I also agree that the solution should be conceptually sound and it should be reasonably obvious how to do the common things. I can see how the model here reflects the ion type system, but I suspect that users will find it confusing, for two reasons: remembering presence or absence of '$' sigil (consider something like Kotlin's '?'), and my experience is that most users, when something is null, don't actually care if it's strongly typed or not.

Were there some other models that were considered and rejected?

Reply:
> Maybe a little refactoring of this section would avoid needing to respecify that

Sure, I can do that.

> Were there some other models that were considered and rejected?

Does this question refer to other models for the built-in types in general, or does it refer specifically to the replacement for `nullable`?

Reply:
Replacement for `nullable`.

Reply:
Yes. Based on the goal of finding a replacement that was explicit about what would be null (ie. no figuring out "base types"), we came up with the following alternatives that were considered:

1. __`nullable` field in type definitions that define which typed nulls are allowed with the nullable annotation.__ This was rejected for being "awkward" and also because it looks like a constraint but behaves unlike any other constraint. The `nullable` field could also be on a type that is imported from another file, making it more difficult for a human reader to discover what sort of null would be allowed.
2. __`nullable::` annotation changed to always allow any sort of null at all.__ This was rejected because it would be weird/unintuitive for `nullable::struct` to allow a `null.int`—it doesn't properly support any use case that people would want to use `nullable` for.
3. __Get rid of nullable entirely__. This was considered the second-best option, but everyone who was consulted agreed that nullability is a common enough use case that we should provide some sort of syntactical sugar for it.
4. __Replace `nullable` with `nullable_text`, `nullable_string`, `nullable_struct`, etc. annotations to indicate which types of nulls should be allowed.__ This was rejected for being more complex than what seemed necessary.  However, rejecting this is not a one-way door with respect to the chosen solution. It would be possible in Ion Schema 2.1 to add annotations like `$null_string_or::`, etc. so that there is syntactical sugar for all of the strongly-typed nulls.